File: packages/oxlint-plugin-react-doctor/src/plugin/rules/jotai/jotai-select-atom-in-render-body.ts
import {
  EFFECT_HOOK_NAMES,
  HANDLER_FUNCTION_NAME_PATTERN,
  HOOK_NAME_PATTERN,
  MEMOIZING_HOOK_NAMES,
  REACT_HANDLER_PROP_PATTERN,
} from "../../constants/react.js";
import { collectHandlerReferencedNames } from "../../utils/collect-handler-referenced-names.js";
import { defineRule } from "../../utils/define-rule.js";
import { findEnclosingFunction } from "../../utils/find-enclosing-function.js";
import {
  isImportedFromModule,
  getImportedNameFromModule,
} from "../../utils/find-import-source-for-name.js";
import { getCalleeName } from "../../utils/get-callee-name.js";
import { getFunctionBindingName } from "../../utils/get-function-binding-name.js";
import type { RuleContext } from "../../utils/rule-context.js";
import type { EsTreeNode } from "../../utils/es-tree-node.js";
import { isFunctionLike } from "../../utils/is-function-like.js";
import { isNodeOfType } from "../../utils/is-node-of-type.js";
import { walkAst } from "../../utils/walk-ast.js";
import type { EsTreeNodeOfType } from "../../utils/es-tree-node-of-type.js";

// HACK: `selectAtom(base, selector)` returns a NEW atom on every call.
// Calling it in a component / hook body without `useMemo` rebuilds
// the derived atom on every render — `useAtomValue` then subscribes
// to a brand new atom each render and triggers an infinite render
// loop (jotai's documented #1 footgun). The fix is either:
//   (a) lift the `selectAtom(...)` call to module scope, or
//   (b) wrap it in `useMemo(() => selectAtom(base, selector), [deps])`.

const JOTAI_SELECT_ATOM_SOURCES = ["jotai/utils", "jotai"];

const COMPONENT_NAME_PATTERN = /^[A-Z]/;

const isImportedSelectAtom = (callExpression: EsTreeNodeOfType<"CallExpression">): boolean => {
  if (!isNodeOfType(callExpression.callee, "Identifier")) return false;
  const localName = callExpression.callee.name;
  for (const source of JOTAI_SELECT_ATOM_SOURCES) {
    if (!isImportedFromModule(callExpression, localName, source)) continue;
    const importedName = getImportedNameFromModule(callExpression, localName, source);
    if (importedName === "selectAtom") return true;
  }
  return false;
};

// A POSITION where `selectAtom(...)` does NOT run on every render: a
// useMemo/useCallback callback (cached), a useEffect/useLayoutEffect callback
// WITH a deps argument (a deps-less effect re-runs after every render —
// still the re-subscribe loop), an inline JSX `onX` attribute value, or an
// `onX` object property.
const isDeferredCallbackPosition = (functionNode: EsTreeNode): boolean => {
  const parent = functionNode.parent;

  if (
    isNodeOfType(parent, "CallExpression") &&
    // First argument only — `useMemo(..., [selectAtomFn])` is not the callback.
    parent.arguments?.[0] === functionNode
  ) {
    // Bare callee (`useMemo(...)`) or namespaced (`React.useMemo(...)`) — read
    // the hook name from the Identifier or the MemberExpression property.
    const hookName = getCalleeName(parent);
    if (hookName && MEMOIZING_HOOK_NAMES.has(hookName)) return true;
    if (hookName && EFFECT_HOOK_NAMES.has(hookName) && Boolean(parent.arguments?.[1])) {
      return true;
    }
  }

  if (isNodeOfType(parent, "JSXExpressionContainer")) {
    const attribute = parent.parent;
    if (
      isNodeOfType(attribute, "JSXAttribute") &&
      isNodeOfType(attribute.name, "JSXIdentifier") &&
      REACT_HANDLER_PROP_PATTERN.test(attribute.name.name)
    ) {
      return true;
    }
  }

  if (isNodeOfType(parent, "Property")) {
    if (
      isNodeOfType(parent.key, "Identifier") &&
      REACT_HANDLER_PROP_PATTERN.test(parent.key.name)
    ) {
      return true;
    }
    if (
      isNodeOfType(parent.key, "Literal") &&
      typeof parent.key.value === "string" &&
      REACT_HANDLER_PROP_PATTERN.test(parent.key.value)
    ) {
      return true;
    }
  }

  return false;
};

const getHandlerNamedBindingName = (functionNode: EsTreeNode): string | null => {
  const parent = functionNode.parent;
  if (
    isNodeOfType(parent, "VariableDeclarator") &&
    isNodeOfType(parent.id, "Identifier") &&
    HANDLER_FUNCTION_NAME_PATTERN.test(parent.id.name)
  ) {
    return parent.id.name;
  }
  if (
    isNodeOfType(functionNode, "FunctionDeclaration") &&
    functionNode.id &&
    HANDLER_FUNCTION_NAME_PATTERN.test(functionNode.id.name)
  ) {
    return functionNode.id.name;
  }
  return null;
};

const containingFunctionIsComponentOrHook = (functionNode: EsTreeNode): boolean => {
  if (isNodeOfType(functionNode, "FunctionDeclaration") && functionNode.id) {
    const declaredName = functionNode.id.name;
    return COMPONENT_NAME_PATTERN.test(declaredName) || HOOK_NAME_PATTERN.test(declaredName);
  }
  // ArrowFunctionExpression / FunctionExpression — look for the
  // surrounding VariableDeclarator. `memo(...)` and `forwardRef(...)`
  // wrappers are transparent here: walk past intermediate calls until
  // we find the binding.
  let cursor: EsTreeNode | null | undefined = functionNode.parent ?? null;
  while (cursor && isNodeOfType(cursor, "CallExpression")) {
    cursor = cursor.parent ?? null;
  }
  if (!cursor) return false;
  if (!isNodeOfType(cursor, "VariableDeclarator")) return false;
  if (!isNodeOfType(cursor.id, "Identifier")) return false;
  return COMPONENT_NAME_PATTERN.test(cursor.id.name) || HOOK_NAME_PATTERN.test(cursor.id.name);
};

// True when `bindingName(...)` is invoked somewhere that executes during
// render — i.e. climbing the invocation's enclosing functions reaches a
// component/hook without crossing a deferred callback or handler-named
// function. A `handle*`/`on*`-named helper that is CALLED on the render path
// still creates a new atom every render, so its name must not suppress.
const isBindingInvokedOnRenderPath = (root: EsTreeNode, bindingName: string): boolean => {
  let didFindRenderPathInvocation = false;
  walkAst(root, (node: EsTreeNode) => {
    if (didFindRenderPathInvocation) return false;
    if (!isNodeOfType(node, "CallExpression")) return;
    if (!isNodeOfType(node.callee, "Identifier") || node.callee.name !== bindingName) return;
    let enclosingFunction = findEnclosingFunction(node);
    while (enclosingFunction) {
      if (
        isDeferredCallbackPosition(enclosingFunction) ||
        getHandlerNamedBindingName(enclosingFunction)
      ) {
        return;
      }
      if (containingFunctionIsComponentOrHook(enclosingFunction)) {
        didFindRenderPathInvocation = true;
        return;
      }
      enclosingFunction = findEnclosingFunction(enclosingFunction);
    }
  });
  return didFindRenderPathInvocation;
};

export const jotaiSelectAtomInRenderBody = defineRule({
  id: "jotai-select-atom-in-render-body",
  title: "selectAtom called during render",
  severity: "error",
  recommendation:
    "Lift `selectAtom(base, fn)` to module scope, or wrap it: `const atom = useMemo(() => selectAtom(base, fn), [deps])`. Calling it during render makes a new atom every time and re-subscribes forever.",
  create: (context: RuleContext) => {
    let cachedRoot: EsTreeNode | null = null;
    const getRoot = (startNode: EsTreeNode): EsTreeNode => {
      if (cachedRoot) return cachedRoot;
      let root: EsTreeNode = startNode;
      while (root.parent) root = root.parent;
      cachedRoot = root;
      return root;
    };

    const isDeferredCallback = (functionNode: EsTreeNode): boolean => {
      if (isDeferredCallbackPosition(functionNode)) return true;
      const handlerName = getHandlerNamedBindingName(functionNode);
      if (!handlerName) return false;
      return !isBindingInvokedOnRenderPath(getRoot(functionNode), handlerName);
    };

    let handlerReferencedNames: Set<string> | null = null;
    const isBindingUsedAsHandler = (functionNode: EsTreeNode): boolean => {
      const bindingName = getFunctionBindingName(functionNode);
      if (!bindingName) return false;
      if (!handlerReferencedNames) {
        handlerReferencedNames = collectHandlerReferencedNames(getRoot(functionNode));
      }
      if (!handlerReferencedNames.has(bindingName)) return false;
      return !isBindingInvokedOnRenderPath(getRoot(functionNode), bindingName);
    };

    return {
      CallExpression(node: EsTreeNodeOfType<"CallExpression">) {
        if (!isImportedSelectAtom(node)) return;

        // Walk up to find the nearest enclosing function. If that
        // function itself is the callback of useMemo / useCallback,
        // the selectAtom call is memoized — fine.
        const nearestFunctionLike = findEnclosingFunction(node);
        if (!nearestFunctionLike) return;
        if (isDeferredCallback(nearestFunctionLike)) return;
        if (isBindingUsedAsHandler(nearestFunctionLike)) return;

        // Now walk up again from the nearest function looking for any
        // enclosing component or hook. Helpers nested inside a
        // component are still "render-time" execution paths.
        let outerCursor: EsTreeNode | null = nearestFunctionLike;
        while (outerCursor) {
          if (isFunctionLike(outerCursor) && containingFunctionIsComponentOrHook(outerCursor)) {
            context.report({
              node,
              message:
                "`selectAtom(...)` runs in a component or hook without `useMemo`, so every render makes a new atom & re-subscribes forever, freezing the page for your users. Lift it to module scope, or wrap it in `useMemo(() => selectAtom(...), [deps])`.",
            });
            return;
          }
          outerCursor = outerCursor.parent ?? null;
        }
      },
    };
  },
});

File: native/oxlint/rules/jotai-select-atom-in-render-body.rs
use oxc_ast::{AstKind, ast::Expression};
use oxc_diagnostics::OxcDiagnostic;
use oxc_macros::declare_oxc_lint;
use oxc_span::GetSpan;

use crate::{AstNode, context::LintContext, rule::Rule};

const MESSAGE: &str = "`selectAtom(...)` runs in a component or hook without `useMemo`, so every render makes a new atom & re-subscribes forever, freezing the page for your users. Lift it to module scope, or wrap it in `useMemo(() => selectAtom(...), [deps])`.";

#[derive(Debug, Default, Clone)]
pub struct JotaiSelectAtomInRenderBody;

declare_oxc_lint!(
    /// Disallow creating a Jotai selectAtom during render.
    JotaiSelectAtomInRenderBody,
    react_doctor_native,
    correctness,
    version = "0.1.0",
    short_description = "selectAtom called during render.",
);

impl Rule for JotaiSelectAtomInRenderBody {
    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
        let AstKind::CallExpression(call) = node.kind() else {
            return;
        };
        let Expression::Identifier(callee) = &call.callee else {
            return;
        };
        let Some(import_entry) = resolve_identifier_import(callee, ctx) else {
            return;
        };
        if !matches!(import_entry.module_request.name(), "jotai" | "jotai/utils")
            || !matches!(
                &import_entry.import_name,
                crate::module_record::ImportImportName::Name(imported_name)
                    if imported_name.name() == "selectAtom"
            )
        {
            return;
        }
        let mut enclosing_functions = ctx
            .nodes()
            .ancestors(node.id())
            .filter(|ancestor| {
                matches!(
                    ancestor.kind(),
                    AstKind::Function(_) | AstKind::ArrowFunctionExpression(_)
                )
            })
            .peekable();
        let Some(nearest_function) = enclosing_functions.peek() else {
            return;
        };
        if jotai_select_atom_is_deferred(nearest_function, ctx)
            || !enclosing_functions
                .any(|function| component_or_hook_function_name(function, ctx).is_some())
        {
            return;
        }
        ctx.diagnostic(OxcDiagnostic::error(MESSAGE).with_label(call.span));
    }
}

fn jotai_select_atom_is_deferred<'a>(function: &AstNode<'a>, ctx: &LintContext<'a>) -> bool {
    let function_root = transparent_expression_root(function, ctx);
    let parent = ctx.nodes().parent_node(function_root.id());
    if let AstKind::CallExpression(call) = parent.kind()
        && call.arguments.first().is_some_and(|argument| {
            argument
                .as_expression()
                .is_some_and(|expression| expression.span() == function_root.span())
        })
        && call.callee_name().is_some_and(|name| {
            matches!(name, "useMemo" | "useCallback")
                || (matches!(name, "useEffect" | "useLayoutEffect" | "useInsertionEffect")
                    && call.arguments.get(1).is_some())
        })
    {
        return true;
    }
    let handler_binding = match function.kind() {
        AstKind::Function(function) => function.id.as_ref().or_else(|| match parent.kind() {
            AstKind::VariableDeclarator(declarator) => declarator.id.get_binding_identifier(),
            _ => None,
        }),
        AstKind::ArrowFunctionExpression(_) => match parent.kind() {
            AstKind::VariableDeclarator(declarator) => declarator.id.get_binding_identifier(),
            _ => None,
        },
        _ => None,
    };
    if let Some(binding) = handler_binding
        && (jotai_handler_name(binding.name.as_str())
            || jotai_binding_is_used_as_handler(binding.symbol_id(), ctx))
        && !jotai_binding_is_invoked_during_render(binding.symbol_id(), ctx)
    {
        return true;
    }
    ctx.nodes()
        .ancestors(function.id())
        .take(4)
        .any(|ancestor| {
            matches!(ancestor.kind(), AstKind::JSXAttribute(attribute)
            if matches!(&attribute.name, oxc_ast::ast::JSXAttributeName::Identifier(identifier)
                if jotai_handler_name(identifier.name.as_str())))
                || matches!(ancestor.kind(), AstKind::ObjectProperty(property)
                if property.key.static_name().is_some_and(|name| jotai_handler_name(&name)))
        })
}

fn jotai_binding_is_used_as_handler(
    symbol_id: oxc_semantic::SymbolId,
    ctx: &LintContext<'_>,
) -> bool {
    ctx.scoping()
        .get_resolved_references(symbol_id)
        .any(|reference| {
            let reference_node = ctx.nodes().get_node(reference.node_id());
            ctx.nodes()
                .ancestors(reference_node.id())
                .take(4)
                .any(|ancestor| {
                    matches!(ancestor.kind(), AstKind::JSXAttribute(attribute)
                        if matches!(&attribute.name, oxc_ast::ast::JSXAttributeName::Identifier(identifier)
                            if jotai_handler_name(identifier.name.as_str())))
                        || matches!(ancestor.kind(), AstKind::ObjectProperty(property)
                            if property.key.static_name().is_some_and(|name| jotai_handler_name(&name)))
                })
        })
}

fn jotai_binding_is_invoked_during_render(
    symbol_id: oxc_semantic::SymbolId,
    ctx: &LintContext<'_>,
) -> bool {
    ctx.scoping()
        .get_resolved_references(symbol_id)
        .any(|reference| {
            let reference_node = ctx.nodes().get_node(reference.node_id());
            let reference_root = transparent_expression_root(reference_node, ctx);
            let parent = ctx.nodes().parent_node(reference_root.id());
            if !matches!(parent.kind(), AstKind::CallExpression(call)
                if call.callee.span() == reference_root.span())
            {
                return false;
            }
            for ancestor in ctx.nodes().ancestors(parent.id()) {
                if !matches!(
                    ancestor.kind(),
                    AstKind::Function(_) | AstKind::ArrowFunctionExpression(_)
                ) {
                    continue;
                }
                if jotai_select_atom_is_directly_deferred(ancestor, ctx) {
                    return false;
                }
                if component_or_hook_function_name(ancestor, ctx).is_some() {
                    return true;
                }
            }
            false
        })
}

fn jotai_select_atom_is_directly_deferred<'a>(
    function: &AstNode<'a>,
    ctx: &LintContext<'a>,
) -> bool {
    let root = transparent_expression_root(function, ctx);
    let parent = ctx.nodes().parent_node(root.id());
    matches!(parent.kind(), AstKind::CallExpression(call)
    if call.arguments.first().is_some_and(|argument| argument.as_expression().is_some_and(|expression| expression.span() == root.span()))
        && call.callee_name().is_some_and(|name| {
            matches!(name, "useMemo" | "useCallback")
                || (matches!(name, "useEffect" | "useLayoutEffect" | "useInsertionEffect")
                    && call.arguments.get(1).is_some())
        }))
        || matches!(parent.kind(), AstKind::VariableDeclarator(declarator)
            if declarator.id.get_binding_identifier().is_some_and(|identifier| jotai_handler_name(identifier.name.as_str())))
        || matches!(function.kind(), AstKind::Function(inner)
            if inner.id.as_ref().is_some_and(|identifier| jotai_handler_name(identifier.name.as_str())))
}

fn jotai_handler_name(name: &str) -> bool {
    name.starts_with("on") && name.as_bytes().get(2).is_some_and(u8::is_ascii_uppercase)
        || name.starts_with("handle") && name.as_bytes().get(6).is_some_and(u8::is_ascii_uppercase)
}
